src/storage-controller/src/lib.rs

```diff
@@ -1694,6 +1747,17 @@ where
             .enumerate()
             .collect::<StreamMap<_, _>>();
 
+        // We cannot check the channel without reading, so try and read, but
+        // then put it back, so that `process()` can work it off.
+        if let Ok(response) = self.read_holds_rx.try_recv() {
+            self.read_holds_tx
+                .send(response)
+                .expect("cannot fail to re-enqueue");
+
+            // Return, to signal readyness for processing!
+            return;
+        }
```

Comment:
Is it actually safe to re-enqueue this way? What if you have two updates for the same read hold? E.g.:

* `(u1, [(100, -1), (200, +1)])`, `(u1, [(200, -1), (300, +1)])`

After this little re-enqueuing stunt, the channel has the following contents:

* `(u1, [(200, -1), (300, +1)])`, `(u1, [(100, -1), (200, +1)])`

I _think_ that's fine based on how the consumer of `read_holds_rx` is implemented, it collects all queued changes first, and only then applies them. But it would be good to have a comment somewhere that this consumer behavior is important for correctness and not just a performance optimization.

Reply:
Yeah, I don't like this stunt, and newer versions of tokio actually have a `len()` and `is_empty()` on channels. I'll add comments that explain what's going on (which you already correctly intuited).